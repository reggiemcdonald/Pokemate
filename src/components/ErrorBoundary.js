import React from "react";
import {
    Text,
    View,
} from "react-native";
import Ionicons from "react-native-vector-icons/Ionicons";
import styles from "../library/styles";

/**
 * **************************
 * A Error FallBack Component
 * **************************
 * Renders when an error is caught
 *
 */
export default class ErrorBoundary extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            hasError: false,
            displayIcon: this.props.displayIcon ? this.props.displayIcon : <Ionicons name={"ios-alert"}/>,
            message: this.props.message ? this.props.message : "Oops! Something happened :("
        }
    }

    static getDerivedStateFromError(error) {
        return {
            hasError: true,
            error: error
        }
    }

    /**
     * Log the error that was caught
     * @override
     * @param error
     * @param errorInfo
     */
    componentDidCatch(error: Error, errorInfo: React.ErrorInfo): void {
        this.setState({
            hasError: true,
            error: error,
            errorInfo: errorInfo
        });
    }

    render() {
        if (this.state.hasError) {
            return(
                <View style={styles.containerCentered}>
                    {this.state.displayIcon}
                    <Text>{this.state.message}</Text>
                </View>
            )
        } else {
            return this.props.children;
        }
    }


}