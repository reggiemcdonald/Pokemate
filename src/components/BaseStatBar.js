import React from "react";
import ErrorBoundary from "./ErrorBoundary";
import {
    View,
    ProgressViewIOS,
    Text
} from "react-native";
import styles, {BaseStatBarLow, BaseStatBarMed, BaseStatBarHigh} from "../library/styles";
import InvalidValue from "../library/errors/InvalidValue";

const MAX_STAT_VALUE = 255;
/**
 * ********************
 * Base Stat Component
 * ********************
 *
 * A react component to display
 * a single base stat of a pokemon
 *
 * ************************************************************
 * As per https://bulbapedia.bulbagarden.net/wiki/Base_stats, Pokemon Base Stats can range from 1 to 255.
 * This component will express the base stat as a proportion of 255
 * props = {
 *     statName: string,
 *     statValue: number between 1 and 255,
 *     [handleChange - implement at a later date]
 * }
 */
export default class BaseStatBar extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            statStyle: this._getStyle()
        }
    }

    // TODO: add the progress bar
    render() {
        return (
            <View>
                <Text>{this.props.statName}</Text>
                {this._renderProgressBar()}
            </View>
        );
    }

    _renderProgressBar() {
        if (this.props.statValue > 255) {
            throw new InvalidValue("Pokemon stat "+this.props.statValue);
        } else {
            const progress = this.props.statValue / MAX_STAT_VALUE;
            return(
                <ProgressViewIOS
                    progress={progress}
                    progressTintColor={this.state.statStyle.progressTint}
                    trackTintColor={this.state.statStyle.trackTint}
                />
            )
        }
    }

    /**
     * Returns the appropriate style according to the statValue
     * @private
     */
    _getStyle() {
        if (this.props.statValue <= 20) {
            return BaseStatBarLow;
        } else if (this.props.statValue > 20 && this.props.statValue < 40) {
            return BaseStatBarMed;
        } else {
            return BaseStatBarHigh;
        }
    }
}