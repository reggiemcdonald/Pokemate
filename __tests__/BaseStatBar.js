import React from "react";
import renderer from "react-test-renderer";
import BaseStatBar from "../src/components/BaseStatBar";
import ErrorBoundary from "../src/components/ErrorBoundary";

describe("Base Stat Bar Render", () => {
    it("BaseStatBar should render without error when using default style", () => {
        // Bulbasaur base HP : 45
        const statName = "HP";
        const statValue = 45;
        const tree = renderer.create(
            <BaseStatBar
                statName={statName}
                statValue={statValue}
            />
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it("BaseStatBar should render ErrorBoundary with invalid value", () => {
        // No value can be greater than 255
        const statName = "HP";
        const statValue = 300;
        const tree = renderer.create(
            <ErrorBoundary>
                <BaseStatBar
                    statName={statName}
                    statValue={statValue}
                />
            </ErrorBoundary>
        );
        expect(tree).toMatchSnapshot()
    });
});