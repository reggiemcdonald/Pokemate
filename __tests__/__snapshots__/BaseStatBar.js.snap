// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base Stat Bar Render BaseStatBar should render ErrorBoundary with invalid value 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#F5FCFF",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    allowFontScaling={false}
    style={
      Array [
        Object {
          "color": undefined,
          "fontSize": 12,
        },
        undefined,
        Object {
          "fontFamily": "Ionicons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    ï„„
  </Text>
  <Text>
    Oops! Something happened :(
  </Text>
</View>
`;

exports[`Base Stat Bar Render BaseStatBar should render without error when using default style 1`] = `
<View>
  <Text>
    HP
  </Text>
  <RCTProgressView
    progress={0.17647058823529413}
    progressTintColor="#00b33c"
    style={
      Array [
        Object {
          "height": 2,
        },
        undefined,
      ]
    }
    trackTintColor="#66ff99"
  />
</View>
`;
